python/src/iceberg/conversions.py

```diff
@@ -0,0 +1,267 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Utility module for various conversions around PrimitiveType implementations
+
+This module enables:
+    - Converting partition strings to built-in python objects
+    - Converting a value to a byte buffer
+    - Converting a byte buffer to a value
+
+Note:
+    Conversion logic varies based on the PrimitiveType implementation. Therefore conversion functions
+    are defined here as generic functions using the @singledispatch decorator. For each PrimitiveType
+    implementation, a concrete function is registered for each generic conversion function. For PrimitiveType
+    implementations that share the same conversion logic, registrations can be stacked.
+
+Example (registering `from_partition_to_py` logic for BooleanType):
+    >>> @from_partition_value_to_py.register(BooleanType)
+    >>> def _(primitive_type, partition_value_as_str: str):
+            if partition_value_as_str is None:
+                return False
+            return partition_value_as_str.lower() == "true"
+
+Example (stacking registrations):
+   >>> @from_partition_value_to_py.register(FloatType)
+   >>> @from_partition_value_to_py.register(DoubleType)
+   >>> def _(primitive_type, partition_value_as_str: str):
+           return float(partition_value_as_str)
+"""
+import struct
+import uuid
+from decimal import ROUND_HALF_DOWN, Decimal, localcontext
+from functools import singledispatch
+
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+
+@singledispatch
+def from_partition_value_to_py(primitive_type, partition_value_as_str: str):
```

Comment:
Should this use `PrimitiveType` for the type argument? Or does that not matter because it uses `@singledispatch`?

Reply:
Hmm, I think in some cases it may not matter but in this case since all of the registered functions inherit from the same superclass I think it's a very useful type-hint. Added it [here](https://github.com/apache/iceberg/pull/4263/files#diff-23863ae95c4894ac0025efabd97dd925503dc5a908e386d7c345952e36b9743cR175).

I also added a type-hint for the `value` argument that describes an aggregate of all of the types possible for the registered functions. Another thing is I noticed I was missing a lot of return type-hints so I added those as well. 👍